### ISSUE TITLE
Feature request: tf.estimator hyperparameter tuning

### ISSUE THREAD

**Issue by carlthome** [tensorflow/tensorflow#16576]
I'm making an issue here because I'm sure this is being worked on somewhere in this huge repo and I've failed to find it by search.

Almost all models have hyperparameters that cannot be set by gradient descent (number of layers, for example). These needs to be tuned, preferably programatically with a smart strategy.

**What's the canonical way of doing hyperparameter tuning with the tf.estimator API?**

(also, how can we do early stopping with tf.estimator?)

I'm currently wrapping around scikit-optimize which is ok, but then I'll never be able to run parallel experiments across workers, and it's a bit tricky to know if the hyperparameters will lead to OOM aside from using tf.profile.

```python
import os

from skopt import gp_minimize
from skopt.space import Real, Categorical, Integer
from skopt.utils import use_named_args

logdir = 'tensorboard/'
space = [
    Real(0.0, 0.1, name='learning_rate'),
    Categorical([True, False], name='skip_connections'),
    Integer(1, 9, name='layers')]


@use_named_args(space)
def score(**params):
    model_dir = os.path.join(logdir, str(params))
    estimator = tf.estimator.Estimator(model_fn, model_dir, params=params)
    trainspec = tf.estimator.TrainSpec(train_input_fn)
    evalspec = tf.estimator.EvalSpec(eval_input_fn)
    try:
        tf.estimator.train_and_evaluate(estimator, trainspec, evalspec)
        metrics = estimator.evaluate(test_input_fn)
        return metrics['loss']
    except (tf.errors.ResourceExhaustedError, tf.train.NanLossDuringTrainingError):
        return 1e9


gp_minimize(score, space)
```

**Comment by drpngx** [comment]
We have been pretty successful with this [approach](https://research.google.com/pubs/pub46180.html) but it's not available publicly. It would be terrific if someone signed up to provide that.

CC @sculleyd

**Comment by carlthome** [comment]
I saw [the presentation on Vizier](https://www.youtube.com/watch?v=12OkQrJvyek) and it looks pretty great but it's unfortunate that it's closed source. A canonical open source alternative would be a very welcome addition to TensorFlow Estimator, even just random search.

I find that particularly getting early stopping and [learning rate annealing as in Keras](https://keras.io/callbacks/#reducelronplateau) into tf.estimator is very messy, as the training and evaluation loop should run distributed so it's not as simple as just having learning rate as a tf.Variable.

**Comment by drpngx** [comment]
Yeah, any implementation in that direction would be useful, I think.

**Comment by BenjaminLarrousse** [comment]
Are you thinking about a package like this one (implemented in R): https://tensorflow.rstudio.com/tools/tfruns/articles/overview.html ?
It would be SO helpful !

**Comment by shoyer** [comment]
It's not open source, but Vizier is usable outside of Google as ["hyperparameter tuning"](https://cloud.google.com/ml-engine/docs/hyperparameter-tuning-overview) in Google Cloud's Machine Learning Engine.

**Comment by hhusum** [comment]
Just chiming in to let you all know that I too would find both early stopping and hyper parameter optimisation highly useful. Particularly early stopping, as hyper parameter optimisation is easier to implement on my own.

**Comment by matthew-z** [comment]
I am also interested in how to do early stopping with Estimator.

**Comment by aroraakshit** [comment]
Very much interested in at least one mechanism for hyper-parameter tuning with Estimator!!

**Comment by carlthome** [comment]
Related: https://github.com/tensorflow/tensorflow/issues/7868

**Comment by carlthome** [comment]
Related: https://github.com/tensorflow/tensorboard/issues/46

**Comment by sushreebarsa** [comment]
@carlthome 
Tf.estimator.Estimator uses Tensorflow 1 style of coding and is not recommended for Tensorflow 2.x. There will not be any new addition to the estimators. 
Could you please follow the detailed explanation [here](https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator) and move this issue to close status? Thanks!

**Comment by carlthome** [comment]
@sushreebarsa sure thing! This has gone quite stale by now.
